tests/anoph/conftest.py

```diff
@@ -584,6 +584,318 @@ def simulate_aim_variants(path, contigs, snp_sites, n_sites_low, n_sites_high):
     return ds
 
 
+def simulate_cnv_hmm(zarr_path, metadata_path, contigs, snp_sites):
+    # zarr_path is the output path to the zarr store
+    # metadata_path is the input path for the sample metadata
+    # contigs is the list of contigs, e.g. Ag has ('2L', '2R', '3R', '3L', 'X')
+    # snp_sites are the data from snp_genotypes/all/sites, inc. {contig}/variants/POS
+    # n_sites will be derived from snp_sites {contig}/variants/POS
+
+    # {release}/cnv/{sample_set}/hmm/zarr
+    # - {contig}
+    #   - calldata
+    #     - CN [2D array] [int] [-1 to 12 for n_sites for n_samples]
+    #     - NormCov [2D array] [float] [0 to 356+ for n_sites for for n_samples]
+    #     - RawCov [2D array] [int] [-1 to 18465+ for n_sites for for n_samples]
+    #   - samples [1D array] [str for n_samples]
+    #   - variants
+    #      - END [1D array] [int for n_snp_sites]
+    #      - POS [1D array] [int for n_snp_sites]
+    # - sample_coverage_variance [1D array] [float] [0 to 0.5 for n_samples]
+    # - sample_is_high_variance [1D array] [bool] [True or False for n_samples]
+    # - samples [1D array] [str]
+
+    # Get a random probability for a sample being high variance, between 0 and 1.
+    p_variance = np.random.random()
+
+    # Open a zarr at the specified path.
+    root = zarr.open(zarr_path, mode="w")
+
+    # Create samples array.
+    df_samples = pd.read_csv(metadata_path)
+    samples = df_samples["sample_id"].values
+    root.create_dataset(name="samples", data=samples, dtype=str)
+
+    # Get the number of samples.
+    n_samples = len(df_samples)
+
+    # Simulate sample_coverage_variance array.
+    sample_coverage_variance = np.random.uniform(low=0, high=0.5, size=n_samples)
+    root.create_dataset(name="sample_coverage_variance", data=sample_coverage_variance)
+
+    # Simulate sample_is_high_variance array.
+    sample_is_high_variance = np.random.choice(
+        [False, True], size=n_samples, p=[1 - p_variance, p_variance]
+    )
+    root.create_dataset(name="sample_is_high_variance", data=sample_is_high_variance)
+
+    for contig in contigs:
+        # Create the contig group.
+        contig_grp = root.require_group(contig)
+
+        # Create the calldata group for this contig.
+        calldata_grp = contig_grp.require_group("calldata")
+
+        # Get the SNP positions for this contig.
+        snp_pos = snp_sites[f"{contig}/variants/POS"][:]
+
+        # Get the number of sites (SNP positions) for this contig.
+        n_sites = snp_pos.shape[0]
```

Comment:
The CNV HMM data is calculated in 300 bp non-overlapping windows across the genome, and is completely independent of the SNP data. I.e., each "variant" is a window. 

The POS array then gives the window start positions, and the END array should be the window end positions (start positions + 299 IIRC). 

So it would be better here to use the length of the contig to derive the number of windows and window start positions.

Reply:
Thanks. Have done.